dds/DCPS/XTypes/DynamicData.cpp

```diff
@@ -577,135 +739,150 @@ DDS::ReturnCode_t DynamicData::get_single_value(ValueType& value, MemberId id,
 {
-  return get_single_value<ACE_CDR::UShort, TK_UINT16>(value, id, TK_BITMASK, 9, 16);
+  return get_single_value<TK_UINT16>(value, id, TK_BITMASK, 9, 16);
 }
 
 DDS::ReturnCode_t DynamicData::get_int64_value(ACE_CDR::LongLong& value, MemberId id)
 {
-  return get_single_value<ACE_CDR::LongLong, TK_INT64>(value, id);
+  return get_single_value<TK_INT64>(value, id);
 }
 
 DDS::ReturnCode_t DynamicData::get_uint64_value(ACE_CDR::ULongLong& value, MemberId id)
 {
-  return get_single_value<ACE_CDR::ULongLong, TK_UINT64>(value, id, TK_BITMASK, 33, 64);
+  return get_single_value<TK_UINT64>(value, id, TK_BITMASK, 33, 64);
 }
 
 DDS::ReturnCode_t DynamicData::get_float32_value(ACE_CDR::Float& value, MemberId id)
 {
-  return get_single_value<ACE_CDR::Float, TK_FLOAT32>(value, id);
+  return get_single_value<TK_FLOAT32>(value, id);
 }
 
 DDS::ReturnCode_t DynamicData::get_float64_value(ACE_CDR::Double& value, MemberId id)
 {
-  return get_single_value<ACE_CDR::Double, TK_FLOAT64>(value, id);
+  return get_single_value<TK_FLOAT64>(value, id);
 }
 
 DDS::ReturnCode_t DynamicData::get_float128_value(ACE_CDR::LongDouble& value, MemberId id)
 {
-  return get_single_value<ACE_CDR::LongDouble, TK_FLOAT128>(value, id);
+  return get_single_value<TK_FLOAT128>(value, id);
 }
 
-DDS::ReturnCode_t DynamicData::get_char8_value(ACE_CDR::Char& value, MemberId id)
-{
-  // String of kind TK_STRING8 is encoded with UTF-8 where each character can take more
-  // than 1 byte. So we can't read a Char8 object, which has size exactly 1 byte,
-  // from such a string as its member.
-  return get_single_value<ACE_CDR::Char, TK_CHAR8>(value, id);
-}
-
-DDS::ReturnCode_t DynamicData::get_char16_value(ACE_CDR::WChar& value, MemberId id)
+template<TypeKind CharKind, TypeKind StringKind, typename ToCharT, typename CharT>
+DDS::ReturnCode_t DynamicData::get_char_common(CharT& value, MemberId id)
 {
   DCPS::Message_Block_Ptr dup(chain_->duplicate());
-  strm_ = DCPS::Serializer(dup.get(), encoding_);
+  setup_stream(dup.get());
 
   const TypeKind tk = type_->get_kind();
   bool good = true;
 
   switch (tk) {
-  case TK_CHAR16:
-    good = (strm_ >> ACE_InputCDR::to_wchar(value));
-    break;
-#ifdef DDS_HAS_WCHAR
-  case TK_STRING16:
+  case CharKind:
+    {
+      ToCharT wrap(value);
+      good = strm_ >> wrap;
+      break;
+    }
+  case StringKind:
     {
-      DCPS::WString str;
+      CharT* str;
       if (!(strm_ >> str)) {
         if (DCPS::DCPS_debug_level >= 1) {
-          ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) DynamicData::get_char16_value -")
+          ACE_ERROR((LM_ERROR, ACE_TEXT("(%P|%t) DynamicData::get_char_common -")
                      ACE_TEXT(" Failed to read wstring with ID %d\n"), id));
         }
         good = false;
         break;
       }
       ACE_CDR::ULong index;
-      if (!get_index_from_id(id, index, str.length())) {
+      const size_t str_len = ACE_OS::strlen(str);
+      if (!get_index_from_id(id, index, str_len)) {
```

Comment:
CI on W19 builds warns about possible loss of value when converting str_len to a ULong.